### ISSUE TITLE
Add option to select the latest possible time with the same low price

### ISSUE THREAD

**Issue by rrooggiieerr** [TheFes/cheapest-energy-hours#110]
Ok, the title is maybe a bit cryptic, but imagine the following use-case:

I want to to prepare the boiler for my morning shower at the cheapest price before I wake up. During the night I have 4 consecutive hours where the price is the same lowest price, actually tonight the price will be € 0,154 from 01:00 till 05:00. The boiler takes max 40 minutes to heat, thus I use the `no_weight_points=3, weight=[2000, 2000]` to configure a runtime of 40 minutes.

My total call looks al follows:
```
{{ as_datetime(cheapest_energy_hours('sensor.nordpool_electricity_price', look_ahead=true, end='07:00', no_weight_points=3, weight=[2000, 2000], mode='end')) }}
```
I run the script when I go to sleep, and when that's before midnight I include `include_tomorrow=true` in the call.

The current implementation gives me the first time the lowest price is valid, thus 01:00, however I like my boiler finished as late as possible, otherwise it's losing some of it's temperature. I like it to start at 04:20 to be ready at 05:00, the end time of the 3 consecutive hours with the same low price.

If I use `mode='end'` I can subtract 40 minutes from the given time. `mode='end'` without any `weight` parameters currently gives me back 02:00, the end time of the first hour with the lowest price. It doesn't give back 05:00, the end time of the last hour with the lowest price.

The following call gives me back 01:20:
```
{{ as_datetime(cheapest_energy_hours('sensor.nordpool_electricity_price', look_ahead=true, end='07:00', include_tomorrow=true, mode='end')) - timedelta(minutes = 40) }}
```
If it would give back 04:20 I'll have exactly what I'm looking for.

Long story short, it would be great if it would be possible to get the latest time possible within the timeframe where same low price applies.

One way to do this would be to introduce a new parameter `latest_possible=True` or implement `mode='end'` to actually give back the end time of the time block with the same low price, 05:00 in the above example.

I'm looking forward to your thoughts on the above, thanks!

**Comment by TheFes** [comment]
Sorry, I don't fully understand what you are trying to achieve with the `no_weight_points` and `weight` combination?
If you want to check for a 45 minutes period, just use `hours=0.75`

But I understand your point. The macro currently returns the first cheapest match it can find. In your case you would like to find the last one.

So assuming a block of 3 hours all with the lowest price, it will return the first 45 minutes, and you would like to have the last 45 minutes. I can have a look at that.

**Comment by rrooggiieerr** [comment]
I totally over looked the `hours=0.75` option, but with the `no_weight_points=3, weight=[2000, 2000]` I indeed tried to achieve the same

**Comment by TheFes** [comment]
well, that will give you 40 minutes, as you split the hour in 3 parts, but you only have 2 weights, so it takes 2/3 of an hour.
It is a bit of a complicated way. And if it is indeed 45 minutes, I would say you need `no_weight_points=4, weight=[1, 1, 1]`
But anyway, `hours=0.75` is a bit easier to use :)

**Comment by rrooggiieerr** [comment]
Sorry for the confusion, but the boiler takes max 40 minutes to heat so ⅔ of an hour is what I was looking for. I'll give the new code a try and report back to you!

**Comment by rrooggiieerr** [comment]
My test results, **bold** are the cases which are not what I expected

I'm testing with the earlier mentioned dataset of € 0,155, € 0,154, € 0,154, € 0,154, € 0,154 and € 0,155 from 00:00 till 06:00 which are the lowest prices of last night.

First, I see the default behaviour of latest_possible is true, which is not backwards compatible and might confuse users after updating:
```
{{ cheapest_energy_hours('sensor.nordpool_electricity_price') }}
```
Returns `2024-02-05T04:00:00+01:00`, **while the previous version returned `2024-02-05T01:00:00+01:00`**

```
{{ cheapest_energy_hours('sensor.nordpool_electricity_price', latest_possible=false) }}
```
Returns `2024-02-05T01:00:00+01:00`, which is the previous behaviour.

```
{{ cheapest_energy_hours('sensor.nordpool_electricity_price', latest_possible=true) }}
```
Returns `2024-02-05T04:00:00+01:00`, which is ok I guess, although there might be some room for debate which behaviour is expected when no time span is known.

Trying to get the earliest possible time for a 45 minutes time span, like the original behaviour:
```
{{ cheapest_energy_hours('sensor.nordpool_electricity_price', hours=0.75, latest_possible=false) }}
```
Returns `2024-02-05T01:00:00+01:00`, which is what it used to be, so that's good

Trying to get the latest possible time:
```
{{ cheapest_energy_hours('sensor.nordpool_electricity_price', hours=0.75, latest_possible=true) }}
```
Returns `2024-02-05T04:15:00+01:00`, which is exactly as expected, so that's good too.

Trying the same 2 use cases, but with a 90 minutes time span:
```
{{ cheapest_energy_hours('sensor.nordpool_electricity_price', hours=1.5, latest_possible=false) }}
{{ cheapest_energy_hours('sensor.nordpool_electricity_price', hours=1.5, latest_possible=true) }}
```
Returns `2024-02-05T01:00:00+01:00` and `2024-02-05T03:30:00+01:00`, which is also exactly as expected.

Trying the same 2 use cases, but with price tolerance of 1 cent
```
{{ cheapest_energy_hours('sensor.nordpool_electricity_price', hours=0.75, latest_possible=false, price_tolerance=0.01) }}
```
Returns `2024-02-05T05:15:00+01:00`, **while I would expect `2024-02-05T00:00:00+01:00`**

```
{{ cheapest_energy_hours('sensor.nordpool_electricity_price', hours=0.75, latest_possible=true, price_tolerance=0.01) }}
```
Retruns `2024-02-05T05:15:00+01:00`, which is exactly as expected, so that's good.

Trying only the price tolerance:
```
{{ cheapest_energy_hours('sensor.nordpool_electricity_price', hours=0.75, price_tolerance=0.01) }}
```
Returns `2024-02-05T05:15:00+01:00` **because the default behaviour of latest_possible is true**

Trying the earlier 2 use cases with `end=true`:
```
{{ cheapest_energy_hours('sensor.nordpool_electricity_price', hours=0.75, mode='end', latest_possible=false) }}
```
Returns `2024-02-05T01:45:00+01:00`, which is exactly as expected, so that's good.

```
{{ cheapest_energy_hours('sensor.nordpool_electricity_price', hours=0.75, mode='end', latest_possible=true) }}
```
Returns `2024-02-05T05:00:00+01:00`, which is exactly as expected, so that's good.

Then the same with a price tolerance of 1 cent:
```
{{ cheapest_energy_hours('sensor.nordpool_electricity_price', hours=0.75, mode='end', latest_possible=false, price_tolerance=0.01) }}
```
Returns `2024-02-05T06:00:00+01:00`, **while I would expect `2024-02-05T00:45:00+01:00`**

```{{ cheapest_energy_hours('sensor.nordpool_electricity_price', hours=0.75, mode='end', latest_possible=true, price_tolerance=0.01) }}
```
Returns `2024-02-05T06:00:00+01:00`, which is exactly as expected

**Comment by rrooggiieerr** [comment]
Regarding the documentation, for `price_tolerance` you mention: _Most useful in combination with `lastest_possible`_

I can think of plenty of uses cases where you want a machine to start as soon as possible within the price tolerance, so I would just leave that out

**Comment by TheFes** [comment]
You're right about the default got `latest_possible`, that should indeed be `false`

`hours` defaults to 1 if no input is given.

I think I implemented the `price_tolerance` differently than you intended, I'll have a look if I can change it

**Comment by rrooggiieerr** [comment]
Does the way I intended `price_tolerance` make sense to you?

**Comment by rrooggiieerr** [comment]
No rush, `latest_possible` is already a great addition!

**Comment by TheFes** [comment]
Could you check again? I think the `price_tolerance` will work more like you expected now.

**Comment by rrooggiieerr** [comment]
@TheFes, yes, all seems to work als expected now! You're the best!

**Comment by TheFes** [comment]
Okay, great.
I need to apply it to the split function as well, so I need some more time before I can publish it. In the meantime you can use this version of course.

**Comment by rrooggiieerr** [comment]
Sorry false alarm, `price_tolerance` seems to work as expected
